### PR TITLE
Fix EZP-27652: Router have wrong context after language switcher link

### PR DESCRIPTION
**Pull request by blankse**
JIRA: https://jira.ez.no/browse/EZP-27652

### CODE REVIEW

**Comment by blankse** [comment]
Review Ping @andrerom @bdunogier @lolautruche

**Comment by lolautruche** [comment]
+1

**Comment by bdunogier** [comment]
@pcardiga this is ready for testing.

File: eZ/Bundle/EzPublishCoreBundle/Routing/DefaultRouter.php [hunk 1]
```diff
@@ -95,7 +96,13 @@ public function generate($name, $parameters = array(), $referenceType = self::AB
...
         }
 
-        $url = parent::generate($name, $parameters, $referenceType);
+        try {
+            $url = parent::generate($name, $parameters, $referenceType);
+        } catch (RouteNotFoundException $e) {
+            // Switch back to original context, for next links generation.
+            $this->setContext($originalContext);
+            throw $e;
+        }
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Routing/DefaultRouter.php]: From the OOP perspective I'd utilize `try-finally` in this case, wrapping the rest of the code like that: https://gist.github.com/alongosz/a6ca06dc34c7d8135503a3ee43751213
It even does not exceed ~100 character in line (more or less), so shouldn't really be too much of a "spaghetti" ;)

Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Routing/DefaultRouter.php]: But then it would have to indent all the code as opposed to return early approach

Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Routing/DefaultRouter.php]: Yes, indentation of the code is a downside. I just realized that we use return early approach in other places, so maybe yes, the code in the form as it is right now, is more consistent.

**Comment by andrerom** [comment]
Merged, nice @blankse 👍

**Pull request merged**